### PR TITLE
fix:duplicate workflow name fixed

### LINKED ISSUE #4868: Bug: Able to create duplicate workflow name in an app. Also not able to clone worflow if workflow name is same
### 📜 Description

Able to create duplicate workflow name in an app. Also not able to clone worflow if workflow name is same .
Screenshot from 2024-04-02 18-20-57

### 👟 Reproduction steps

.

### 👍 Expected behavior

should not allow to create workflow with an existing workflow name

### 👎 Actual Behavior

able to create workflow with an existing workflow name

### ☸ Kubernetes version

.

### Cloud provider

<details>

</details>

### 🌍 Browser

Chrome

### 🧱 Your Environment

_No response_

### ✅ Proposed Solution

_No response_

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find any similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/devtron-labs/devtron/blob/main/CODE_OF_CONDUCT.md)

AB#9240

### PR DESCRIPTION
**Pull request by Laeeqdev**
# Description

We've addressed a critical issue where users could inadvertently create duplicate workflow names within an app, leading to confusion and potential errors. Additionally, users were unable to clone workflows if their names were identical, limiting workflow management flexibility. With our recent fixes, the system now prevents duplicate workflow names within an app and allows cloning workflows with identical names across different apps, enhancing usability and efficiency.
Fixes #4868

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: api/restHandler/CoreAppRestHandler.go [hunk 1]
```diff
@@ -1587,6 +1587,16 @@ func (handler CoreAppRestHandlerImpl) createWorkflows(ctx context.Context, appId
...
 func (handler CoreAppRestHandlerImpl) createWorkflowInDb(workflowName string, appId int, userId int32) (int, error) {
+	//checking if workflow name is already exist or not
+	workflow, err := handler.appWorkflowRepository.FindByNameAndAppId(workflowName, appId)
+	if err != nil && !errors.Is(err, pg.ErrNoRows) && !errors.Is(err, pg.ErrMultiRows) {
+		handler.logger.Errorw("error in finding workflow by app id and name", "name", workflowName, "appId", appId)
+		return 0, err
+	}
+	// if workflow name already exists then we will assign a new name to the workflow
+	if workflow.Id != 0 {
+		workflowName = fmt.Sprintf("%s-clone-%s", workflowName, util.Generate(4))
```
Comment (prakash100198) [inline api/restHandler/CoreAppRestHandler.go:1598]: make a func for this, maybe something GenerateNewWorkflowName, also use this everywhere

**Review by prakash100198 — CHANGES_REQUESTED** [review]

**Review by prakash100198 — APPROVED** [review]
lgtm

File: api/restHandler/CoreAppRestHandler.go [hunk 2]
```diff
@@ -1587,6 +1587,16 @@ func (handler CoreAppRestHandlerImpl) createWorkflows(ctx context.Context, appId
 }
 
 func (handler CoreAppRestHandlerImpl) createWorkflowInDb(workflowName string, appId int, userId int32) (int, error) {
+	//checking if workflow name is already exist or not
+	workflow, err := handler.appWorkflowRepository.FindByNameAndAppId(workflowName, appId)
+	if err != nil && !errors.Is(err, pg.ErrNoRows) && !errors.Is(err, pg.ErrMultiRows) {
```
Comment (prakash100198) [inline api/restHandler/CoreAppRestHandler.go:1592]: take this code in appworkflowservice and return if workflow exist or not

File: pkg/appWorkflow/AppWorkflowService.go [hunk 3]
```diff
@@ -178,7 +179,11 @@ func (impl AppWorkflowServiceImpl) CreateAppWorkflow(req AppWorkflowDto) (AppWor
 	var wf *appWorkflow.AppWorkflow
 	var savedAppWf *appWorkflow.AppWorkflow
 	var err error
-
+	workflow, err := impl.appWorkflowRepository.FindByNameAndAppId(req.Name, req.AppId)
```
Comment (prakash100198) [inline pkg/appWorkflow/AppWorkflowService.go:182]: make a func checkIfWorkflowExists func for checking if workflow exists or not

**Review by prakash100198 — CHANGES_REQUESTED** [review]

**Review by prakash100198 — DISMISSED** [review]

**Review by prakash100198 — APPROVED** [review]

**Review by kripanshdevtron — APPROVED** [review]

**Review by prakash100198 — APPROVED** [review]

**Pull request merged**